File: ruby/lib/scl_parser.rb
# frozen_string_literal: true

require_relative "scl/errors"
require_relative "scl/lexer"
require_relative "scl/parser"
require_relative "scl/serializer"

module SCL
  VERSION = "0.1.0"
  AUTHOR = "shareui"

  module_function

  def loads(text)
    tokens = Lexer.tokenize(text)
    Parser.parse(tokens)
  end

  def load(path, encoding: "UTF-8")
    loads(File.read(path, mode: "r:bom|utf-8", encoding: encoding))
  end

  def dumps(hash, indent: 4)
    Serializer.new(indent: indent).serialize(hash)
  end

  def dump(hash, path, indent: 4, encoding: "UTF-8")
    File.write(path, dumps(hash, indent: indent), mode: "w:utf-8", encoding: encoding)
  end
end

File: elixir/scl_parser/lib/scl_parser.ex
defmodule SCL do
  @moduledoc """
  SCL parser for Elixir.

  API matches Python version:
    - load/1, loads/1
    - dump/2, dumps/1
  """

  alias SCL.Parser.Lexer
  alias SCL.Parser.Core
  alias SCL.Serializer

  @doc "Parse string"
  @spec loads(String.t()) :: map()
  def loads(text) when is_binary(text) do
    tokens = Lexer.tokenize(text)
    Core.parse(tokens)
  end

  @doc "Parse file"
  @spec load(Path.t()) :: map()
  def load(path) when is_binary(path) do
    path |> File.read!() |> loads()
  end

  @doc "Map → string"
  @spec dumps(map()) :: String.t()
  def dumps(data) when is_map(data) do
    Serializer.serialize(data)
  end

  @doc "Map → file"
  @spec dump(map(), Path.t()) :: :ok
  def dump(data, path) when is_map(data) and is_binary(path) do
    path |> File.write!(dumps(data))
    :ok
  end
end